### ISSUE TITLE
Combine LeaderSkill.AREA_CHOICES and Dungeon.CATEGORY_CHOICES

### ISSUE THREAD

**Issue by claytondaley** [swarfarm/swarfarm#98]
I'm trying to figure out how to determine whether a `LeaderSkill` applies in a specific `Level` (i.e. dungeon instance).  The `LeaderSkill` stores the area in an `IntegerField` (i.e. `LeaderSkill.area`) that references the local `AREA_CHOICES` object.  `Dungeon` (i.e. `Level.dungeon`) has a separate, local `Dungeon.category` that draws from CATEGORY_CHOICES.

If both were FKs to the same `Model`, it would be much easier to make this determination.

**Comment by PeteAndersen** [comment]
How would you propose leader skills that apply to a specific element monster?

**Comment by claytondaley** [comment]
`LeaderSkill.element` already uses `Monster.ELEMENT_CHOICES` (`None` if it doesn't apply).  Similarly, `LeaderSkill.area` should reference a common set of areas with `Dungeon` or maybe the more abstract `Source` (`None` if it doesn't apply).  The current `AREA_GENERAL` becomes implicit as `element is None and area is None`.

**Comment by claytondaley** [comment]
Someone could plan a team for any of the categories covered by the skill restrictions -- Siege, Guild War, Arena, Dungeons-- plus Scenarios (in theory not necessarily practice), Raid, and TOA.  I don't think the current values for `Dungeon` include Arena/Guild content, but it seems the obvious extension.

It's not immediately obvious to me where you would make the `Dungeon` vs `Level` distinction, but there are plenty of reasonable ways.  It probably makes sense to treat GW, Siege, and Lab each as a `Dungeon` with a category of `Guild`.  These could contain `Level` values like GWO, GWD, 4* Siege D, 5* Siege D, 4* Siege O, 5* Siege O, and various Lab bosses/types.  Similarly, `Arena` (`Dungeon`) would have O and D (and possibly more granularity).

**Comment by PeteAndersen** [comment]
Yes, I had planned on including PVP areas into the dungeons eventually pretty much as you described. I'll go ahead and add the categories to the Dungeon model now, but probably not add zones just yet. 

The logic for determining if a leader skill applies to a dungeon/level is better off being implemented external to these two models. There's no good union of the options for both models. LeaderSkills boil down to PvE, Arena, or Guild for the dungeon types. Dungeons have many more specific PvE categories. With any given leader skill and any Dungeon instance it should be easy enough to check if the dungeon is arena or guild, else it's PvE.

**Comment by claytondaley** [comment]
The "Dungeon" `LeaderSkill` only applies to Cairos and Rift... not TOA and Scenario... so you can't just exclude Arena and Guild.  Obviously, you *can* do a manual mapping (and I will, at least during dev).

At the same time, it seems to me that this is a property of the `Dungeon`.  Perhaps it does not replace `Dungeon.category` but is instead an additional `Dungeon.area`.  I really *want* to merge them because the values in `category` appear to be specializations (i.e. subsets) of the `area` (at least the way I want to define it).  Logically, they're part of one tree (though that may not be the parsimonious design decision).

I'm sure my motivation is also partly philosophical.  I don't like implicit FKs in a DB (in this case, FKs to a non-existent table).  If two fields refer to the same value set, I like to materialize the value set and make the FK relationship explicit.  You can always do `a.element_id == b.element_id` to avoid the database lookup, but the relationship is available to be traversed at the DB if you ever had a reason to.

(It occurs to me that the swarfarm UI is not friendly to explicit FKs.  For example, the custom category logic in views don't work unless they're a stable, internally defined constant.  If this wasn't the case, explicit FKs also benefit from being extensible.  Adding a new dungeon category in the DB would create a URL, automatically populate select boxes, etc.)

**Comment by claytondaley** [comment]
While we're on the topic of dungeon types, do you track SDs?  Do they fall into an existing category?  Like Scenario, I doubt anyone would explicitly specify a team, but it's another category of dungeon that could theoretically need to be cross-referenced with leader skills.

**Comment by PeteAndersen** [comment]
I'll add a `Dungeon.area` and having LeaderSkill reference that.

The UI is in life support mode, I am putting zero further work into the Django views, templates, and URL routing and the only interface to the Django application will be the API. The API provides this information without explicit tables and rows of data via OPTIONS requests.

**Comment by claytondaley** [comment]
Thanks!

Have you updated the fixture(s) to include whatever Dungeon(s)/Level(s) you've created?

**Comment by PeteAndersen** [comment]
I will probably not update the fixture itself until the last moment, but the dungeon data is all being parsed from data files which are included in the repo. The functions for parsing were added here: https://github.com/PeteAndersen/swarfarm/blob/logs_redux/bestiary/com2us_data_parser.py#L774

I forgot to answer the question about secret dungeons - I do not plan on logging since the drops are fixed and well known, but they will be added under the dungeons section probably with their own category, and a 'general' area.
